transformers4rec/torch/model/base.py

```diff
@@ -543,52 +523,64 @@ def __init__(
         self.head_weights = head_weights or [1.0] * len(head)
         self.head_reduction = head_reduction
         self.optimizer = optimizer
-        self.hf_format = hf_format
 
-    def forward(self, inputs: TensorOrTabularData, training=True, **kwargs):
+    def forward(
+        self, inputs: TensorOrTabularData, training=True, testing=False, targets=None, **kwargs
+    ):
         # TODO: Optimize this
         outputs = {}
-        for head in self.heads:
-            outputs.update(
-                head(inputs, call_body=True, training=training, always_output_dict=True, **kwargs)
-            )
-
-        if len(outputs) == 1:
-            outputs = outputs[list(outputs.keys())[0]]
-            if isinstance(outputs, dict):
-                # next-item prediction task with `hf_format = True``  returns a dictionary
-                # with three tensors: loss, predictions, labels. This needed for training
-                # and evaluation using the Trainer class.
-                # At inference, we just need the predictions tensors.
-                # TODO: We are simplifying the logic around `hf_format` in the multi-gpu
-                # support work.
-                if not self.hf_format:
-                    return outputs["predictions"]
-            return outputs
 
-        return outputs
-
-    def compute_loss(self, inputs, targets, compute_metrics=True, **kwargs) -> torch.Tensor:
-        losses = []
-
-        for i, head in enumerate(self.heads):
-            loss = head.compute_loss(
-                inputs, targets, call_body=True, compute_metrics=compute_metrics, **kwargs
-            )
-            losses.append(loss * self.head_weights[i])
+        if training or testing:
+            losses = []
+            labels = {}
+            predictions = {}
+            for i, head in enumerate(self.heads):
+                head_output = head(
+                    inputs, call_body=True, training=training, testing=testing, **kwargs
+                )
+                labels.update(head_output["labels"])
+                predictions.update(head_output["predictions"])
+                losses.append(head_output["loss"] * self.head_weights[i])
+            loss_tensor = torch.stack(losses)
+            loss = getattr(loss_tensor, self.head_reduction)()
+            if len(labels) == 1:
+                labels = list(labels.values())[0]
+                predictions = list(predictions.values())[0]
+            outputs = {"loss": loss, "labels": labels, "predictions": predictions}
 
-        loss_tensor = torch.stack(losses)
+        else:
+            for head in self.heads:
+                outputs.update(
+                    head(inputs, call_body=True, training=training, testing=testing, **kwargs)
+                )
+                if len(outputs) == 1:
```

Comment:
This condition should happen outside of the for loop otherwise we will always return the outputs of the first head.